### ISSUE TITLE
save calibration parameter (mle_model_results.MLEResults)

### ISSUE THREAD

**Issue by fgilbert01** [statsmodels/statsmodels#2867]
Hello,

I am using SARIMAX model to forecast hourly electric load. When, I used 1 year of hourly historical data, the fit of the model is long so I want to be able to save in a file my SARIMAX model results to be able to forecast in a later session.
I would like to know if it is possible to save the model results that will only contain the optimized calibration parameter. The file created by "sarimax_model_results.save(cur_w_dir + OUTPUT_FILE_SARIMAX_PARAMETER, remove_data=False)" is 600 MByte large. I would like to know if we can make it smaller.

**Comment by josef-pkt** [comment]
The intention was for `remove_data=True` to remove large memory attributes that are not needed for prediction. I'm not sure that works for time series models, when we need at least some previous fitted observations to continue with the forecasting.

@ChadFulton   Did you ever look at this for the statespace or other timeseries models?

My guess is that to enable forecasting we would need to keep at least maxlag observations and residuals.

**Comment by josef-pkt** [comment]
@fgilbert01   Are you using the forecast at the end of the estimation dataset, or are you forecasting for a new dataset?

**Comment by fgilbert01** [comment]
I have tried to set "remove_data=False" but got a problem after when I ran the prediction "sarimax_model_results.predict(start= 0, end= iend_index, exog= None)". The error is : "object of type 'NoneType' has no len()".
I am forecasting at the end of the estimated dataset.

**Comment by ChadFulton** [comment]
This feature is not well supported in state space models, and I'm somewhat surprised that it works at all. It probably shouldn't be too much trouble to make it work in some fashion, but the main question for me is how this function is intended to work / how we want it to work.

@fgilbert01 how are you hoping this feature would work, both in the `remove_data=False` and `remove_data=True` cases?

@josef-pkt how was this feature intended to work in Statsmodels generally?

**Comment by josef-pkt** [comment]
I assume you meant "True".

I don't think we had a discussion about this before. So, I kind of expected that we either don't remove enough or too much. Chad will know better the details for SARIMAX.
In general you might be better off storing the forecast itself, given that there is only one use of forecasting.

I see three possible solutions
- keep a copy (to avoid a view of the entire array) of the last observations, and check that forecast is indexing from the end and not based on the full array index. This might be quite a bit of work for a pretty limited usecase.
- Use the predict helper function given the stored params and user or internally stored observations for previous observations and residuals.
- Use a filter to filter an entire time series to get the residuals for an MA part, and use for forecasting.

specific to statespace models:
The statespace representation at the last observation should include the `state` that is needed for forecasting, so it should be possible to just store the statespace instance with the current/last state.

**Comment by josef-pkt** [comment]
@ChadFulton  The usecase for pickling with remove_data was for prediction when we have independent observations. In that case params, plus possibly cov_params, provides all the information to make new predictions. 
Given that we don't have standalone predict functions, this was a relatively easy way for users to be able to pickle and store a model and use it for prediction on new datasets.

The problems is that we need more information than just the params for these kinds of models, e.g. with dependence as in time series models, or kernel regression or some nonparametric methods would need the full data, and nobody ever checked what the `remove_data` pickle version should be.
(Some models like random effects models only provide prediction for the fixed effects part, but in time series models we don't want to ignore the effect coming from the estimated residuals.)

The usecase for time series models that would be useful to support is to estimate the model once and then use it for filtering or predicting new datasets without necessarily keeping the original estimation dataset around.

**Comment by fgilbert01** [comment]
My use case is the following: I want to make an electrical load forecast on a raspberry pi computer that has a very weak processor. To save processing time, I thought I would calibrate the SARIMAX model only once a month and use the latter most recent calibration parameter when I run my forecast every hour. I though the calibration parameter could be stored in a separate file. 

This way, I would fit the model with a large set of observation data but forecast with a smaller set of observation data. This can be done with hydrological model but I do not know if this apply with time series model.

**Comment by josef-pkt** [comment]
@fgilbert01  thanks for explaining the usecase.

It's an interesting usecase, and we had similar questions before (e.g. recalibrate/estimate only in fixed intervals and reuse existing parameter estimates for immediate updating of the forecasts which is also useful if estimation takes more time and forecasts need to be available very fast.). The main problem is that it is not trivial to get the required information that is needed for the forecast for new datasets.
Your case is more difficult than just doing a single forecast at the end because you also need to update the information during the month when you have obtained more information. So, it's closer to dynamic prediction but on a new sample of observations.

**Comment by ChadFulton** [comment]
Great, this is just what I was looking for. It looks like we need to support three cases:

1. Fit the model, save the results, and later do pure forecasting.
2. Fit the model, save the results, and later add new data to the same sample, then do forecasting.
3. Fit the model, save the results, and later use the fitted parameters on an entirely new dataset.

In the current save / load functionality in Statsmodels is there a way to re-attach data to a loaded object? Is there an example somewhere?

**Comment by josef-pkt** [comment]
> In the current save / load functionality in Statsmodels is there a way to re-attach data to a loaded object? 

No, this question never came up before.

Unless you have already pre-processed data, then it is difficult to attach arrays.
For example predict (outside of tsa) has some processing for exog to get the right format (and patsy transformation if necessary).
It is also model specific which arrays that are attributes have been removed (keywords for attributes to try to remove, are attached to model and results, which can include weights, offset, exposure and other extra arrays, e.g. also wexog, wendog in WLS that are creating during model instantiation.).

One generic possibility would be to create a new model which would have the default data processing and then copy (reassign nor copy memory) the relevant attributes arrays to the `old` model and results instances.

Another possibility would be to attach the new model to the old results instance (`results._results.model = new_model`), but that is more fragile because additional attributes like degrees of freedom would be wrong for the original estimation.

**Comment by fgilbert01** [comment]
Hello Josef,

What would be the next step for the issue above. Have you agreed that this is something you will make available in the statsmodels package. If yes, do you know when this will be available.

**Comment by josef-pkt** [comment]
@fgilbert01  We keep the issue open, because we need support for this.

I don't know how soon something like this will be available. Nobody is currently working on the ARMA/ARIMA models, and SARIMAX depends on Chad.

I expect that a partial remove_data will be a bit tricky and will require many unit tests.
I think a standalone function or class that just forecasts based on a stored state would be easier to implement, if the objective is prediction. If standard errors or prediction errors are also needed then it gets a bit more complicated.
The current code is a bit complex because of all the datetime handling for specifying forecast and prediction periods. I guess that a limited, 1 or h period ahead forecast without datetimes would be easier.

**Comment by josef-pkt** [comment]
see also #2788 for the same usecase (independent of pickle)

**Comment by josef-pkt** [comment]
@ChadFulton  How difficult would it be to write a class that just keeps track of the state (statespace instance) and calculates a few updating steps for the forecast?

Can the statespace instance be separated from the estimated model/results, or is there too much integration?

**Comment by ChadFulton** [comment]
It shouldn't be difficult to solve this problem. The state space is a little more flexible than maybe other types of models because it already explicitly has a method for applying a specific set of parameters to a model / dataset through the `filter` / `smooth` methods.

Basically there are only two things that need to be saved to do any of the three things I mentioned above:

- The parameters of the model
- The last elements of the `predicted_state` and `predicted_state_cov` arrays (which are all that is required to save the "state" of the filter).

Below I have sketched out how each of the three scenarios above would work. The main issue here is in the interface - determining exactly what data / results are saved / deleted, and how the `load` and `save` functions would work

(The below code *almost* works, though there's currently a bug preventing filtering with nobs=0; there's nothing technically preventing filtering / smoothing in this case, it's just a time index check that always raises an exception if nobs=0).

```python
import numpy as np
import pandas as pd
import statsmodels.api as sm

dta = sm.datasets.macrodata.load_pandas().data
dta.index = pd.date_range(start='1959-01-01', end='2009-07-01', freq='QS')

initial_sample = dta.realgdp.ix[:'1999-10-01']
later_sample = dta.realgdp.ix['1999-10-01':]
np.random.seed(1234)
unrelated_sample = np.random.normal(size=10)

# Estimate the model on the initial sample
est_mod = sm.tsa.SARIMAX(initial_sample, order=(4,1,1))
est_res = est_mod.fit()
print est_res.forecast(2)
```

    2000-01-01    11139.517087
    2000-04-01    11253.080554
    Freq: QS-JAN, dtype: float64

```python
# Save the parameters and the predicted state values
est_params = est_res.params
est_state = est_res.predicted_state[:, -1]
est_state_cov = est_res.predicted_state_cov[:, :, -1]
```

> (1) Fit the model, save the results, and later do pure forecasting.

```python
# Sometime later you want to forecast some values...
pure_forecast_mod = sm.tsa.SARIMAX([], order=(4,1,1))
pure_forecast_mod.initialize_known(est_state, est_state_cov)
# This is the part that doesn't currently quite work because there's a bug
# where an exception is raised when nobs=0. There is a simple fix though.
pure_forecast_res = pure_forecast_mod.smooth(est_params)
print pure_forecast_res.forecast(2)
```

    [ 11139.51708714  11253.08055365]

> (2) Fit the model, save the results, and later add new data to the same sample, then do forecasting.

```python
# Sometime later you get new data, but don't want to refit...
later_mod = sm.tsa.SARIMAX(later_sample, order=(4,1,1))
later_mod.initialize_known(est_state, est_state_cov)
later_res = later_mod.smooth(est_params)
print later_res.forecast(2)
```

    2009-10-01    13035.207979
    2010-01-01    13083.874956
    Freq: QS-JAN, dtype: float64

> (3) Fit the model, save the results, and later use the fitted parameters on an entirely new dataset.

```python
# Sometime later you get a new dataset, but thing the parameters are the same
unrelated_mod = sm.tsa.SARIMAX(unrelated_sample, order=(4,1,1))
unrelated_res = unrelated_mod.smooth(est_params)
print unrelated_res.forecast(2)
```

    [-3.04737337 -3.80746906]

**Comment by josef-pkt** [comment]
@ChadFulton  that looks good.
Can the est_state be pickled? how large is that?

@fgilbert01  Can you be more specific how your predict workflow works, or is supposed to work?

I would start by wrapping one of Chad's version into a new class that just pickles est_params, est_state, est_state_cov and has update_data and forecast methods.
This would provide a minimal working version that can be expanded as needed.

**Comment by ChadFulton** [comment]
> Can the est_state be pickled? how large is that?

Definitely. `est_state` and `est_state_cov` are very small, `k_states x 1` and `k_states x k_states`, where `k_states` depends on the model but is usually very small (in the above example `k_states = 5`).

**Comment by josef-pkt** [comment]
One question for feature of this: Do we forecast only once, and create a new SARIMAX instance each time? Or, do we need `dynamic` prediction?

The former would be a standard usecase for online prediction without reestimation.
The latter would be useful and faster for backtesting, i.e. make the new data large but then predict dynamically starting at each time period.

**Comment by fgilbert01** [comment]
The use case is the following:

I am using a year of hourly historical electric load to fit a SARIMAX model. From the last value of the historical load, I forecast for the next 24 hours. I run the forecast every hour but would like to be able to fit the model only once a week because I want the forecasting to go fast. The model is run in an operational mode.

If I could fit the SARIMAX model with a year of historical data. At hour=1, I would fit and save my SARIMAX model parameter separately in a file. For the hour=2,3,etc, I would only get a certain amount of historical data, maybe 4 weeks of historical data and re-run the SARIMAX for the next 24 hours using the SARIMAX calibration parameter that I have saved in a file just after my run at hour=1. N.B. I do not know if it is necessary to load any historical values before predicting.

**Comment by josef-pkt** [comment]
> For the hour=2,3,etc, I would only get a certain amount of historical data, maybe 4 weeks of historical data and re-run the SARIMAX for the next 24 hours using the SARIMAX calibration parameter that I have saved in a file just after my run at hour=1. N.B. I do not know if it is necessary to load any historical values before predicting.

To see if I understand correctly:
First SARIMAX (if you have an MA term) needs the estimated residuals before the prediction date. 
You can estimate them again by filtering from some previous periods (weeks). The length of the data that would be good to have depends on the persistence of the ARMA, i.e. coefficients/roots close to 1. (that is estimate the residuals but not the parameters)

The alternative is to store the residuals at the last estimation, or updating. In this case the `state` contains all the necessary historical information. In this case the new state after a new observation arrives needs to be stored, so it can be used as the basis for the next prediction.
In Chads example code above, version 1, this might be initializing a new model with one observation (instead of empty) and given stored prior state.

**Comment by ChadFulton** [comment]
I'm not sure I'm tracking this properly. But there is no need to include historical data or residuals if you're going to use the same parameters and you use the appropriate elements of the `predicted_state` and `predicted_state_cov` matrices.

**Comment by fgilbert01** [comment]
@josef-pkt I think you understood well.
The residual should also be saved in a separate file and updated once we got new observation data. I suppose residual file is linked to only one ARIMAX model parameter. How long the residual file should be can tricky but the programmer can make a test to run a forecast with a short and a long residual to see if makes a big difference on the forecast.

**Comment by ChadFulton** [comment]
What are you planning to do with the residuals?

**Comment by josef-pkt** [comment]
@ChadFulton   If you have the state available and save it then you don't need the residuals. However, predict in ARMA uses past observations plus past estimated residuals for prediction. There is no saved `state`. 
Also, when we have a completely new dataset, then we would need to filter the entire time series to get the current `state` or residuals for the forecast/prediction. (i.e. `filter` with given params)

**Comment by ChadFulton** [comment]
Okay, that makes sense. I thought you were going to use the residuals in the state space context somehow.

**Comment by fgilbert01** [comment]
Chad, Joseph, you were quick to answer.
I can make myself available to test the changes you have discussed above with my case. Can you let me know when theses changes would be available. I could make the test with a specific development version and give my feedback.

**Comment by fgilbert01** [comment]
@ChadFulton

Chad, I think I understood what I need to do with scenario #2. My observed electrical load data have a 24 hour season. Then I would do the following:

- Fit the SARIMAX model with 8760 observed values (1 year)
- Save the 'predicted_state'  and  'predicted_state_cov' parameter each one in a separate file
- Run the forecast and terminate execution
- When I receive a new observation data, I need to make a new forecast so:
- I create a new instance of the SARIMAX model but takes only the last 7 days of data =  168 values. The 168 values contains the new value
- later_mod = sm.tsa.SARIMAX( [168values],order=(1,0,0),seasonal_order=(0,1,1,24))
-  I retrieve the state parameter that was saved in the a separate files and assign:
- later_mod.initialize_known(est_state, est_state_cov)
- I call the smooth command later_res = later_mod.smooth(est_params)
- And after run the forecast

This way I do not have the problem with the nobs=0 and can use the statsmodels library as is. The run should be fast because I have fewer observation data in the second run.

**Comment by ChadFulton** [comment]
Yes, that looks right to me.

**Comment by fgilbert01** [comment]
Josef, Chad,

I have followed your instruction and the workflow to save the SARIMAX calibration parameter is working just fine. Thank you for your quick answer. I recall having issues with software from large companies like ESRI and the support was not that quick.

As far as I am concerned, the case can be closed.

**Comment by josef-pkt** [comment]
@fgilbert01  The case is still open for other users.

It would be great if you could post your solution in a gist or notebook as recipe, or as a PR for eventual inclusion.
IMO we will have to support this directly, at least to get the required helper functions and methods for the models.

(we are open source)

**Comment by fgilbert01** [comment]
The solution is following what Chad has written above.
Can you explain me how Chad proceed to write explanation with source code embedded. I will make the explanation with the same format.

**Comment by vinner54** [comment]
Hi I have a similar task to what was discussed above, but I want to update my data very frequently and always just add one additional data point. It works using case 2 or 3 as described by Chad Fulton. However, the whole smoothing procedure is costly in computation time (too costly for my aim), thus I would like to do just a manual updating of the statespace vector with the newly arriving data point. Any advice ?
